components/kyma-environment-broker/internal/process/deprovisioning/avs_evaluations.go

```diff
@@ -0,0 +1,51 @@
+package deprovisioning
+
+import (
+	"time"
+
+	"github.com/kyma-incubator/compass/components/kyma-environment-broker/internal"
+	"github.com/kyma-incubator/compass/components/kyma-environment-broker/internal/avs"
+	"github.com/kyma-incubator/compass/components/kyma-environment-broker/internal/storage"
+	"github.com/sirupsen/logrus"
+)
+
+type AvsEvaluationRemovalStep struct {
+	delegator             *avs.Delegator
+	operationsStorage     storage.Operations
+	externalEvalAssistant avs.EvalAssistant
+	internalEvalAssistant avs.EvalAssistant
+}
+
+func NewAvsEvaluationsRemovalStep(delegator *avs.Delegator, operationsStorage storage.Operations, externalEvalAssistant, internalEvalAssistant avs.EvalAssistant) *AvsEvaluationRemovalStep {
+	return &AvsEvaluationRemovalStep{
+		delegator:             delegator,
+		operationsStorage:     operationsStorage,
+		externalEvalAssistant: externalEvalAssistant,
+		internalEvalAssistant: internalEvalAssistant,
+	}
+}
+
+func (ars *AvsEvaluationRemovalStep) Name() string {
+	return "De-provision_AVS_Evaluations"
+}
+
+func (ars *AvsEvaluationRemovalStep) Run(deProvisioningOperation internal.DeprovisioningOperation, logger logrus.FieldLogger) (internal.DeprovisioningOperation, time.Duration, error) {
+	if deProvisioningOperation.AVSExternalEvaluationDeleted && deProvisioningOperation.AVSInternalEvaluationDeleted {
+		logger.Infof("Both internal and external evaluations have been deleted")
+		return deProvisioningOperation, 0, nil
+	}
+
+	provisioningOperation, err := ars.operationsStorage.GetProvisioningOperationByInstanceID(deProvisioningOperation.InstanceID)
+	if err != nil {
+		logger.Errorf("error while getting provisioning deProvisioningOperation from storage")
+		return deProvisioningOperation, time.Second * 10, nil
```

Comment:
We don't want to fail or timeout the de-provisioning operation if some step is constantly failing. So our idea is to implement a function in the `internal/process/deprovision_operation.go`:
```
// RetryOperationWithoutFail retries an operation for at maxTime in retryInterval steps and omits the operation if retrying failed
func (om *DeprovisionOperationManager) RetryOperationWithoutFail(operation internal.DeprovisioningOperation, description string, retryInterval time.Duration, maxTime time.Duration, log logrus.FieldLogger) (internal.DeprovisioningOperation, time.Duration, error) {
	since := time.Since(operation.UpdatedAt)

	log.Infof("Retrying for %s in %s steps", maxTime.String(), retryInterval.String())
	if since < maxTime {
		return operation, retryInterval, nil
	}
	// update description to track failed steps
	updatedOperation, repeat := om.update(operation, domain.InProgress, description)
	if repeat != 0 {
		return updatedOperation, repeat, nil
	}

	log.Errorf("Omitting after %s of failing retries", maxTime.String())
	return updatedOperation, 0, nil
}
```
Using this function, you can execute a given number of retries for the error paths. In case your step exceeds the timeout, the deprovisioning operation won't fail and the processing queue will move to the next step.

Reply:
@polskikiel , it would be nice if all these are conventions are the part of the documentation.
The implementation so far was based on the documentation https://github.com/kyma-incubator/compass/blob/master/docs/kyma-environment-broker/03-03-runtime-provisioning-and-deprovisioning.md.

This could save time and reduce unnecessary wasted effort.

Reply:
@abbi-gaurav You're right, I will update the documentation and add such a function in separate PR. Then you can just rebase and use it in your step.

Reply:
https://github.com/kyma-incubator/compass/pull/1164